src/library/scala/collection/immutable/ListSet.scala

```diff
@@ -174,9 +170,15 @@ sealed class ListSet[A] extends AbstractSet[A]
 
     /** `-` can be used to remove a single element from a set.
      */
-    override def -(e: A): ListSet[A] = if (e == head) self else {
-      val tail = self - e; new tail.Node(head)
-    }
+    override def -(e: A): ListSet[A] = removeInternal(e, this, Nil)
+
+    @tailrec private def removeInternal(k: A, cur: ListSet[A], acc: List[ListSet[A]]): ListSet[A] =
+      if (cur.isEmpty)
+        acc.last
```

Comment:
These lines are pretty short.  It's easier to follow the logic if the result is on the same line as the conditional.  (You already went this way with the `hasNext` thing above, which is more debatable.)